File: ios/Classes/Shaper/NetworkConditionShaper.swift
import Foundation

/// Shared shaper contract matching Android NetworkConditionShaper.
public final class NetworkConditionShaper {
  private var latencyMs: Double = 0
  private var jitterMs: Double = 0
  private var packetLoss: Double = 0
  private var offline: Bool = false
  private var uploadBytesPerSecond: Double = .infinity
  private var downloadBytesPerSecond: Double = .infinity
  private let lock = NSLock()

  public init() {}

  public func update(config: TunnelConfig) {
    lock.lock()
    defer { lock.unlock() }
    latencyMs = config.latencyMs
    jitterMs = config.jitterMs
    packetLoss = config.packetLoss
    offline = config.isOffline
    uploadBytesPerSecond = mbpsToBps(config.uploadMbps)
    downloadBytesPerSecond = mbpsToBps(config.downloadMbps)
  }

  public func shouldDrop() -> Bool {
    lock.lock()
    defer { lock.unlock() }
    if offline { return true }
    if packetLoss <= 0 { return false }
    return Double.random(in: 0..<1) < min(max(packetLoss, 0), 1)
  }

  public func shape(byteCount: Int) {
    lock.lock()
    let latency = latencyMs
    let jitter = jitterMs
    let rate = downloadBytesPerSecond
    lock.unlock()

    if rate.isFinite && rate > 0 {
      let seconds = Double(byteCount) / rate
      Thread.sleep(forTimeInterval: seconds)
    }
    var delay = latency
    if jitter > 0 {
      delay += Double.random(in: -jitter...jitter)
    }
    if delay > 0 {
      Thread.sleep(forTimeInterval: delay / 1000.0)
    }
  }

  private func mbpsToBps(_ mbps: Double) -> Double {
    if !mbps.isFinite { return .infinity }
    if mbps <= 0 { return 0 }
    return mbps * 1024 * 1024 / 8
  }
}

File: android/src/main/kotlin/com/darkmintis/network_simulator/shaper/NetworkConditionShaper.kt
package com.darkmintis.network_simulator.shaper

import com.darkmintis.network_simulator.config.TunnelConfig

/**
 * Contract for applying real network conditions to packet/byte flows.
 */
interface TrafficShaper {
    fun updateConfig(config: TunnelConfig)

    /** Returns false when the unit must be dropped (loss/offline). */
    fun admit(direction: TrafficDirection, sizeBytes: Int): Boolean

    /** Blocks until bandwidth + latency/jitter allow transmission. */
    fun shape(direction: TrafficDirection, sizeBytes: Int)
}

class NetworkConditionShaper : TrafficShaper {
    private val delayLoss = DelayLossModel()
    private val uploadBucket = TokenBucket(Double.POSITIVE_INFINITY)
    private val downloadBucket = TokenBucket(Double.POSITIVE_INFINITY)

    @Volatile
    private var config: TunnelConfig = TunnelConfig()

    override fun updateConfig(config: TunnelConfig) {
        this.config = config
        delayLoss.latencyMs = config.latencyMs
        delayLoss.jitterMs = config.jitterMs
        delayLoss.packetLoss = config.packetLoss
        delayLoss.offline = config.isOffline
        uploadBucket.updateRate(mbpsToBytesPerSecond(config.uploadMbps))
        downloadBucket.updateRate(mbpsToBytesPerSecond(config.downloadMbps))
    }

    override fun admit(direction: TrafficDirection, sizeBytes: Int): Boolean {
        if (config.isOffline) return false
        return !delayLoss.shouldDrop()
    }

    override fun shape(direction: TrafficDirection, sizeBytes: Int) {
        val bucket = if (direction == TrafficDirection.UPLOAD) uploadBucket else downloadBucket
        while (!bucket.tryConsume(sizeBytes)) {
            val wait = bucket.waitTimeMillis(sizeBytes)
            interruptibleSleep(wait.coerceAtMost(50))
        }
        val delay = delayLoss.nextDelayMillis()
        if (delay > 0) {
            interruptibleSleep(delay)
        }
    }

    private fun interruptibleSleep(millis: Long) {
        try {
            Thread.sleep(millis)
        } catch (_: InterruptedException) {
            Thread.currentThread().interrupt()
        }
    }

    private fun mbpsToBytesPerSecond(mbps: Double): Double {
        if (!mbps.isFinite()) return Double.POSITIVE_INFINITY
        if (mbps <= 0) return 0.0
        return mbps * 1024.0 * 1024.0 / 8.0
    }
}
